x/yarpctest/request_unary.go

```diff
@@ -58,11 +58,25 @@ func HTTPRequest(options ...api.RequestOption) api.Action {
 		require.NoError(t, out.Start())
 		defer func() { assert.NoError(t, out.Stop()) }()
 
-		resp, cancel, err := sendRequest(out, opts.GiveRequest, opts.GiveTimeout)
-		defer cancel()
-		validateError(t, err, opts.WantError)
-		if opts.WantError == nil {
-			validateResponse(t, resp, opts.WantResponse)
+		for i := 0; i < opts.RetryCount+1; i++ {
+			resp, cancel, err := sendRequest(out, opts.GiveRequest, opts.GiveTimeout)
+			defer cancel()
```

Comment:
defer in a loop queues the cancel until the function returns, not the end of scope. This should either be called manually or you can scope the request inside the loop with a closure.

Reply:
Wrapped the request in a func closure so that I can use defer there.